core/opregion.c

```diff
@@ -38,571 +41,327 @@ void lai_write_opregion(lai_nsnode_t *field, lai_variable_t *source) {
-                bit_offset = field->fld_offset % 8;
-                break;
-
-            case FIELD_WORD_ACCESS:
-                bit_offset = field->fld_offset % 16;
-                offset &= (~1);        // clear lowest bit
-                break;
-
-            case FIELD_DWORD_ACCESS:
-            case FIELD_ANY_ACCESS:
-                bit_offset = field->fld_offset % 32;
-                offset &= (~3);        // clear lowest two bits
-                break;
+        progress += 8;
+    }
+}
 
-            case FIELD_QWORD_ACCESS:
-                bit_offset = field->fld_offset % 64;
-                offset &= (~7);        // clear lowest three bits
-                break;
+// this assumes little endian
+static uint64_t lai_buffer_read_at(uint8_t *buffer, size_t byte_offset, size_t num_bytes) {
+    uint64_t value = 0;
 
-            default:
-                lai_panic("undefined field flags 0x%02x: %s", field->fld_flags,
-                          lai_stringify_node_path(field));
-        }
-    } else {
-        bit_offset = field->fld_offset % 32;
-        pci_byte_offset = field->fld_offset % 4;
+    for (size_t i = 0; i < num_bytes; i++) {
+        value |= (uint64_t)buffer[i] << (i << 3);
     }
 
-    if(opregion->op_override){
-        if(instance->trace & LAI_TRACE_IO)
-            lai_debug("lai_read_field: Reading from overrided field with address space %02X, address: %X", opregion->op_address_space, opregion->op_base + offset);
-        switch (field->fld_flags & 0x0F) {
-        case FIELD_BYTE_ACCESS:
-            if(!opregion->op_override->readb)
-                lai_panic("Opregion override doesn't provide readb function");
-            value = opregion->op_override->readb(opregion->op_base + offset, opregion->op_userptr) >> bit_offset;
-            break;
+    return value;
+}
 
-        case FIELD_WORD_ACCESS:
-            if(!opregion->op_override->readw)
-                lai_panic("Opregion override doesn't provide readw function");
-            value = opregion->op_override->readw(opregion->op_base + offset, opregion->op_userptr) >> bit_offset;
-            break;
 
-        case FIELD_DWORD_ACCESS:
-        case FIELD_ANY_ACCESS:
-            if(!opregion->op_override->readd)
-                lai_panic("Opregion override doesn't provide readd function");
-            value = opregion->op_override->readd(opregion->op_base + offset, opregion->op_userptr) >> bit_offset;
-            break;
+static size_t lai_calculate_access_width(lai_nsnode_t *field) {
+    lai_nsnode_t *opregion = field->fld_region_node;
+    
+    size_t access_size;
+    switch (field->fld_flags & 0xF) {
+        case FIELD_BYTE_ACCESS: access_size = 8; break;
+        case FIELD_WORD_ACCESS: access_size = 16; break;
+        case FIELD_DWORD_ACCESS: access_size = 32; break;
+        case FIELD_QWORD_ACCESS: access_size = 64; break;
+        case FIELD_ANY_ACCESS: {
+            access_size = 1 << (32 - __builtin_clz(field->fld_size - 1));
```

Comment:
Add a comment what this does (round up to next power of 2). We might want to `_Static_assert` that `int` is 32-bit wide.